src/main/scala/chisel3/util/SparseVec.scala

```diff
@@ -0,0 +1,230 @@
+
+  // Determine if a default value needs to exist.  This should only exist if
+  // there are fewer map items than the size of the Vec.  (As this is a
+  // SparseVec, this is expected to be a Some.)
+  private val defaultValue: Option[A] = mapping match {
+    case full if mapping.size == size => None
+    case _                            => Some(WireInit(gen, DontCare))
+  }
+
+  /** Read a value from a [[SparseVec]] using one of several possible lookup
+    * typs. The returned value is read-only.
+    *
+    * @param addr the address of the value to read from the vec
+    * @param lookupType the type of lookup, e.g., binary, one-hot, or when-based
+    * @param sourceinfo implicit source locator information
+    * @return a read-only value from the specified address
+    * @throws ChiselException if the returned value is written to
+    */
+  def apply(addr: UInt, lookupType: SparseVec.Type = SparseVec.Binary)(implicit sourceinfo: SourceInfo): A = {
+    val result: A = lookupType match {
+      // Generate a lookup using a decoder.
+      case d: SparseVec.Decoder =>
+        // The number of bits required to represent all addresses of the vector.
+        val addrWidth = log2Up(size)
+
+        // The address where user-provided data starts.  If there is no default
+        // value needed, then start user data at address zero.
+        val baseEncodedAddress = defaultValue match {
+          case Some(_) => 1
+          case None    => 0
+        }
+
+        // The total number of output addresses that need to be encoded.
+        val encodedSize = elements.size + baseEncodedAddress + {
+          zeroValue match {
+            case Some(_) => 1
+            case None    => 0
+          }
+        }
+
+        val legalBitPats = mapping.zipWithIndex.map {
+          case (index, i) =>
+            BitPat(index.U(addrWidth.W)) -> BitPat(d.encoding(i + baseEncodedAddress, encodedSize))
+        } ++ {
+          // If out-of-bounds behavior is possible, then add BitPats for all
+          // out-of-bounds accesses.
+          zeroValue match {
+            case None => Seq.empty
+            case Some(_) =>
+              (size until BigInt(addrWidth).pow(2).toInt).map {
+                case i =>
+                  BitPat(i.U(addrWidth.W)) -> BitPat(d.encoding(baseEncodedAddress, encodedSize))
+              }
+          }
+        }
+
+        // Generate the truth table.  Use a default value of zero.
+        val ttable = TruthTable(
+          legalBitPats,
+          BitPat.N(encodedSize)
+        )
+
+        // Generate the lookup.
+        d.lookup(decoder(addr, ttable), VecInit((defaultValue ++ elements.values).toSeq)).asInstanceOf[A]
+
+      // Generate a lookup using a sequence of when statements.
+      case SparseVec.IfElse =>
+        val result = Wire(gen)
+
+        // The default value _must_ be specified to bypass initialization checking.
+        result := 0.U
```

Comment:
The user may want to opt-in to DontCare here.

Reply:
I think that's reasonable as an option. This is behavior that is different from the `DenseVec` case. E.g., making this change here will cause the tests to fail because the default value is now the first element (or however a FIRRTL compiler decides to compile this).

Reply:
Adding a bunch of stuff to allow this.